simpleviewinc/tap-watchtower
# ZEN-551: Tap Watchtower

**Ticket**: [ZEN-551](https://jira.simpleviewtools.com/browse/ZEN-551)
> When I created this repo, I committed a few files copied from keg-herkin as its first commit on master, so that’s why you will see the file diffs referring to keg-herkin things. You can just ignore those diffs -- everything should be reviewed fresh.

## Context
* we want to run the watchtower container on the keg staging server to automatically restart containers when a new image becomes available

## Goal

* setup a tap with custom cli tasks for managing the watchtower container

## Updates

* `configs/watchtower-template.config.js`
  * this is the template for a watchtower configuration
  * the `start` task will look for this in the caller's home directory. If it is present, it will use its values for the `start` parameters, unless the user also defined command line parameters (which take precedence)
* `tasks/definitions/container/attach.js`
  * override attach task that simply attaches to the watchtower container
* `tasks/definitions/container/kill.js`
  * stops the container, with a `force` option
* `tasks/definitions/container/status.js`
  * simple task that prints out the status of the container and the current resolved configuration
* `tasks/definitions/container/start.js`
  * starts the watchtower container
  * accepts many different flags, which are also configurable by the `watchtower.config.js(on)` filed
* `tasks/utils/getConfig.js`
  * gets the config object merged from all config sources (custom path, home directory, configs/ directory, etc.)

## Testing

* clone `tap-watchtower` into your `keg-hub/taps` directory
* `yarn install`
*  `keg tap link watchtower` (or whatever alias you want)
* `yarn test`
  * tests should pass
* create a configuration file in your home directory (use the template in `configs/watchtower-template.config.js`)
  * set `runOnce` property to `true` in yours
* `keg watchtower start --att --debug`
  * verify that it runs the container once, then immediately exits (and removes the container)
* `keg watchtower start --no-once --debug`
  * this should run the container indefinitely in debug mode
* `keg watchtower stat`
  * verify the container is running
* `keg watchtower att`
  * verify that you can attach to the container 
  * `ctrl-p ctrl-q` to exit
* `keg watchtower kill`
  * verify that it kills the container
